matlab/src/cpp/arrow/matlab/array/proxy/numeric_array.h

```diff
@@ -0,0 +1,93 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include "arrow/array.h"
+#include "arrow/builder.h"
+#include "arrow/type_traits.h"
+
+#include "arrow/matlab/array/proxy/array.h"
+
+#include "libmexclass/proxy/Proxy.h"
+
+namespace arrow::matlab::array::proxy {
+
+template<typename CType>
+class NumericArray : public arrow::matlab::array::proxy::Array {
+    public:
+        NumericArray(const libmexclass::proxy::FunctionArguments& constructor_arguments)
+            : arrow::matlab::array::proxy::Array(constructor_arguments) {
+            using ArrowType = typename arrow::CTypeTraits<CType>::ArrowType;
+            using BuilderType = typename arrow::CTypeTraits<CType>::BuilderType;
+
+            // Get the mxArray from constructor arguments
+            const ::matlab::data::TypedArray<CType> numeric_mda = constructor_arguments[0];
+            const ::matlab::data::TypedArray<bool> make_copy = constructor_arguments[1];
+
+            // Get raw pointer of mxArray
+            auto it(numeric_mda.cbegin());
+            auto dt = it.operator->();
+
+            const bool make_deep_copy = make_copy[0];
+
+            if (make_deep_copy) {
+                BuilderType builder;
+                auto st = builder.AppendValues(dt, numeric_mda.getNumberOfElements());
+
+                // TODO: handle error case
+                if (st.ok()) {
+                    auto maybe_array = builder.Finish();
+                    if (maybe_array.ok()) {
+                        array = *maybe_array;
+                    }
+                }
+            } else {
+                // Pass pointer to Arrow array constructor that takes a buffer
+                // Do not make a copy when creating arrow::Buffer
+                std::shared_ptr<arrow::Buffer> buffer(
+                                                      new arrow::Buffer(reinterpret_cast<const uint8_t*>(dt),
+                                                                        sizeof(CType) * numeric_mda.getNumberOfElements()));
+                
+                // Construct arrow::NumericArray specialization using arrow::Buffer.
+                // pass in nulls information...we could compute and provide the number of nulls here too
+                std::shared_ptr<arrow::Array> array_wrapper(
+                                                            new arrow::NumericArray<ArrowType>(numeric_mda.getNumberOfElements(), buffer,
+                                                                                               nullptr, // TODO: fill validity bitmap with data
+                                                                                               -1));
```

Comment:
How about creating an `ArrayData` https://github.com/apache/arrow/blob/main/cpp/src/arrow/array/data.h and then creating an `Array` by `MakeArray()` https://github.com/apache/arrow/blob/main/cpp/src/arrow/array/util.h#L34-L38 ?

We may be able to share the same implementation with other array types with this approach.